### ISSUE TITLE
[FR] Improve MLflow CI times & reduce test flakiness

### ISSUE THREAD

**Issue by smurching** [mlflow/mlflow#2344]
## Overview
This probably resonates with many of you - as we've added new functionality to MLflow, CI times have grown, and unfortunately [test flakiness](https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html) has increased (largely within the Python build). Moreover, the MLflow master build has broken increasingly frequently as the set of test dependencies has grown. We should consider ways to improve this & make the MLflow development loop faster & more stable for everyone. Your thoughts & ideas are welcome here!

## Goals
To start the discussion, I'd propose that we define a couple of key metrics & aim to improve on them (welcome to other ideas here):
1. **Pass rate of CI builds on the `master` branch**: when PRs are merged into MLflow, we trigger a run of all tests against the master branch. The pass rate of these builds relates closely to test flakiness, in that we only merge PRs that pass all tests, so a master-build failure indicates that a test passed on a PR and then failed when merged into master (is flaky).
2. **Average open-source PR build time**: we should aim to reduce this without compromising on test coverage
3. **Number of master build breakages** : this differs from 1 & refers specifically to the master build being deterministically broken (e.g. most often this occurs due to changes or breakages in test dependencies).

### Pass rate of CI builds on the `master` branch
A first step would be to measure this - we can get a rough idea of build pass rate via [Travis CI insights](https://travis-ci.org/mlflow?tab=insights), but cannot view pass rates per build. A more involved alternative would be to export test results to a third-party aggregator & view summary stats there. On Travis CI, we might achieve this through use of [Build Artifacts](https://docs.travis-ci.com/user/uploading-artifacts/) (i.e. export the results of master builds as artifacts in a format like jUnit XML). Other CI tools (e.g. [Circle CI](https://circleci.com/docs/2.0/collect-test-data/)) allow for the same flow, but simplified (& provides native visualization of test failures/pass rates etc).

Once we can measure & identify tests that are failing in master, we can take action to fix them, and temporarily mitigate the impact of their flakiness via tools like https://github.com/box/flaky (for Python tests at least).

### Average open-source PR build time
Historically, we've attempted to improve build times by grouping tests into stages & that run serially (failures of any tests in earlier stages prevents execution of later stages), e.g. gated execution of large integration tests on smaller unit tests passing. 

Most MLflow PRs involve well-scoped changes that affect only a limited set of functionality (this is a good thing), but we nonetheless run every unit/integration test on each push. We would therefore likely see significant reductions from the use of selective test execution like that provided by [testmon](https://testmon.org/) for Python, i.e. only running tests corresponding to changed code. We can still run the full set of tests on merges to master as a failsafe, but other large projects (e.g. [conda](https://github.com/tarpas/pytest-testmon/network/dependents?dependents_after=NjQwNjg1OTk4OQ&package_id=UGFja2FnZS01MjUwMzA3Ng%3D%3D))  seem to successfully depend on testmon.

### Number of master build breakages
Entirely preventing breakages caused by changes in third-party dependencies is difficult unless we remove said dependencies. However, selective test execution as discussed above could reduce this (by simply not executing broken tests). Similarly, we could pull per-framework modules out of MLflow (e.g. pull the `mlflow.tensorflow` module into its own package & test/maintain it independently of MLflow core). 

## Proposal
Tackle low-hanging fruit first & migrate to Circle CI in order to address longer-term issues/challenges of visibility into test pass rates, etc. Circle CI also seems to have a [cleaner config format](https://circleci.com/docs/2.0/sample-config/#sample-configuration-with-sequential-workflow) (i.e. no "default" build stage), other nice features like support for [custom docker images](https://circleci.com/docs/2.0/custom-images/), & is [used by other OSS projects](https://circleci.com/open-source/) like React, PyTorch, etc

#### Short Term
* (~2 person-days) Investigate selective test execution via testmon to speed CI times
* (~1 person-day / ongoing) Use https://github.com/box/flaky & proactively label flaky tests to reduce their impact on build success/failure until they're fixed
* (~4 days) Investigate ways to cache MLflow test environment for faster & more robust env setup, e.g:
  * Build a docker image containing MLflow test dependencies & configure CircleCI to update this image (push to DockerHub) on merges. Should make sure we consider if & how this complicates the process of updating or adding new test dependencies in the process. Will also need to figure out how to make this work with builds that run against non-master branches (do they all run against the same test docker image?).
  * Use [Circle CI's build caching](https://circleci.com/docs/2.0/caching/) to cache the conda installation directory
#### Medium-term
* (~2 person-days) Start migrating to Circle CI, by first enabling it alongside Travis CI & only merging PRs when Travis passes. Requires refactoring Travis build logic to run on Circle CI

#### Longer-term:
* (~1 person day) Switch over to Circle CI, once we have confidence in its stability & speed.
* (~3+ person days) Fix all flaky tests (once we switch to Circle CI, it should be easier to identify which tests are in fact flaky)

### Use other CI systems/tools
Would love to hear suggestions/recommendations on other CI systems or tools that'd help achieve our goals.

### Continue using Travis CI
One alternative would be to stick with Travis CI - in this world, the proposal above might change to become:

#### Short Term
* (~2 person-days) Investigate selective test execution via testmon to speed CI times
* (~1 person-day / ongoing) Use https://github.com/box/flaky & proactively label flaky tests to reduce their impact on build success/failure until they're fixed
* (~4 days) Investigate ways to cache MLflow test environment for faster & more robust env setup, e.g:
  * Build a docker image containing MLflow test dependencies & configure Travis to update this image (push to DockerHub) on merges. Should make sure we consider if & how this complicates the process of updating or adding new test dependencies in the process. Will also need to figure out how to make this work with builds that run against non-master branches (do they all run against the same test docker image?).
  * Use [Travis CI's build caching](https://docs.travis-ci.com/user/caching/) to cache the conda installation directory

#### Medium Term
* (~1 person-days) Use Travis build artifacts to upload test reports to S3
* (? person-days) Configure a third-party service to consume test reports & provide summary statistics on test pass rates

#### Longer Term
* (~3+ person days) Fix all flaky tests

**Comment by smurching** [comment]
BTW, thinking we can actually get started on fixes that are independent (or close to it) of our choice of CI tool for now - in particular:

* (~1 person-day / ongoing) Use https://github.com/box/flaky & proactively label flaky tests to reduce their impact on build success/failure until they're fixed. Given that we don't have per-test pass-rate reporting yet, one way to get started here is to look through failed master builds [here](https://travis-ci.org/mlflow/mlflow/builds), identify unit tests that failed recently on master, and label them as flaky.
* (~2 person-days) Investigate selective test execution via testmon to speed CI times. The challenge here seems like it'll be finding a way to ensure testmon's cache files (.testmondata) are updated when PRs are merged. We can look for inspiration at what other projects that use testmon (e.g. modin) do - it looks like modin for example (see [related PRs](https://github.com/modin-project/modin/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+testmon) & [umbrella issue](https://github.com/modin-project/modin/issues/853)) uses Github workflows to upload the .testmondata file to S3 on merges to master & fetches it at the start of PR builds.

Let me know if you'd be interested in either of these!

**Comment by harupy** [comment]
@smurching Can we close this issue because we have completely migrated to GitHub Actions?

**Comment by smurching** [comment]
@harupy yes! Thanks for all the hard work on this!
